controllers/storagecluster/reconcile.go

```diff
@@ -355,9 +356,42 @@ func (r *StorageClusterReconciler) reconcilePhases(
 
 	// Check GetDeletionTimestamp to determine if the object is under deletion
 	if instance.GetDeletionTimestamp().IsZero() {
+		updateRequired := false
 		if !contains(instance.GetFinalizers(), storageClusterFinalizer) {
 			r.Log.Info("Finalizer not found for StorageCluster. Adding finalizer.", "StorageCluster", klog.KRef(instance.Namespace, instance.Name))
 			instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, storageClusterFinalizer)
+			updateRequired = true
+		}
+
+		if instance.Spec.AllowRemoteStorageConsumers {
+			updateRequired = util.AddAnnotation(
+				instance,
+				util.BackwardCompatabilityInfoAnnotationKey,
+				string(util.JsonMustMarshal(util.BackwardCompatabilityInfo{Pre4_19DeploymentMode: util.ProviderDeploymentMode})),
+			)
+		} else if instance.Spec.ExternalStorage.Enable {
+			updateRequired = util.AddAnnotation(
+				instance,
+				util.BackwardCompatabilityInfoAnnotationKey,
+				string(util.JsonMustMarshal(util.BackwardCompatabilityInfo{Pre4_19DeploymentMode: util.ExternalDeploymentMode})),
+			)
+		} else {
+			svgPre4_19 := &rookCephv1.CephFilesystemSubVolumeGroup{}
+			svgPre4_19.Name = fmt.Sprintf("%s-%s", util.GenerateNameForCephFilesystem(instance.Name), subVolumeGroupName)
+			svgPre4_19.Namespace = instance.Namespace
+			if err := r.Get(r.ctx, client.ObjectKeyFromObject(svgPre4_19), svgPre4_19); client.IgnoreNotFound(err) != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to get pre4_19 subvolumegroup cr: %v", err)
+			}
+			if svgPre4_19.UID != "" {
+				updateRequired = util.AddAnnotation(
+					instance,
+					util.BackwardCompatabilityInfoAnnotationKey,
+					string(util.JsonMustMarshal(util.BackwardCompatabilityInfo{Pre4_19DeploymentMode: util.InternalDeploymentMode})),
+				)
+			}
+		}
```

Comment:
Can you please separate the calculation of the backward compatibility struct into a separate function?